core/requests/trie_access.go

```diff
@@ -0,0 +1,200 @@
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/core/access"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/trie"
+	"golang.org/x/net/context"
+)
+
+// TrieAccess implements trie.OdrAccess, providing database/network access for
+// a trie identified by root hash
+type TrieAccess struct {
+	trie.OdrAccess
+	ca     *access.ChainAccess
+	root   common.Hash
+	trieDb trie.Database
+}
+
+// NewTrieAccess creates a new TrieAccess
+func NewTrieAccess(ca *access.ChainAccess, root common.Hash, trieDb trie.Database) *TrieAccess {
+	return &TrieAccess{
+		ca:     ca,
+		root:   root,
+		trieDb: trieDb,
+	}
+}
+
+// RetrieveKey retrieves a single key, returns true and stores nodes in local
+// database if successful
+func (self *TrieAccess) RetrieveKey(ctx context.Context, key []byte) bool {
+	//fmt.Println("request trie %v key %v", self.root, key)
+	r := NewTrieEntryAccess(self.root, self.trieDb, key)
+	return self.ca.Retrieve(ctx, r) == nil
+}
+
+// OdrEnabled returns true if this TrieAccess is capable of doing network requests
+func (self *TrieAccess) OdrEnabled() bool {
+	return self.ca.OdrEnabled()
+}
+
+// TrieEntryAccess is the ODR request type for state/storage trie entries
+type TrieEntryAccess struct {
+	access.ObjectAccess
+	root       common.Hash
+	trieDb     trie.Database
+	key, value []byte
+	proof      trie.MerkleProof
+	skipLevels int // set by DbGet() if unsuccessful
+}
+
+// NewTrieEntryAccess creates a new TrieEntryAccess request
+func NewTrieEntryAccess(root common.Hash, trieDb trie.Database, key []byte) *TrieEntryAccess {
+	return &TrieEntryAccess{root: root, trieDb: trieDb, key: key}
+}
+
+// Request sends an ODR request to the LES network (implementation of access.ObjectAccess)
+func (self *TrieEntryAccess) Request(peer *access.Peer) error {
+	glog.V(access.LogLevel).Infof("ODR: requesting trie root %08x key %08x from peer %v", self.root[:4], self.key[:4], peer.Id())
+	req := &access.ProofReq{
+		Root: self.root,
+		Key:  self.key,
+	}
+	return peer.GetProofs([]*access.ProofReq{req})
+}
+
+// Valid processes an ODR request reply message from the LES network
+// returns true and stores results in memory if the message was a valid reply
+// to the request (implementation of access.ObjectAccess)
+func (self *TrieEntryAccess) Valid(msg *access.Msg) bool {
+	glog.V(access.LogLevel).Infof("ODR: validating trie root %08x key %08x", self.root[:4], self.key[:4])
+
+	if msg.MsgType != access.MsgProofs {
+		glog.V(access.LogLevel).Infof("ODR: invalid message type")
+		return false
+	}
+	proofs := msg.Obj.([]trie.MerkleProof)
+	if len(proofs) != 1 {
+		glog.V(access.LogLevel).Infof("ODR: invalid number of entries: %d", len(proofs))
+		return false
+	}
+	value, err := trie.VerifyProof(self.root, self.key, proofs[0])
+	if err != nil {
+		glog.V(access.LogLevel).Infof("ODR: merkle proof verification error: %v", err)
+		return false
+	}
+	self.proof = proofs[0]
```

Comment:
it is thread safe to write the results because Deliver that calls this lock chainAccess including all requests sent. Please comment that caller must hold the lock over req to be used concurrently. 
the same for Valid functions of all request types